mozilla_django_oidc/auth.py

```diff
@@ -123,52 +153,52 @@ def authenticate(self, **kwargs):
         }
 
         # Get the token
-        response = requests.post(self.OIDC_OP_TOKEN_ENDPOINT,
-                                 data=token_payload,
-                                 verify=import_from_settings('OIDC_VERIFY_SSL', True))
-        response.raise_for_status()
+        token_response = self.get_token_response(token_payload)
+        id_token = token_response.get('id_token')
+        access_token = token_response.get('access_token')
 
         # Validate the token
-        token_response = response.json()
-        id_token = token_response.get('id_token')
-        if self.verify_token(id_token, nonce=nonce):
-            access_token = token_response.get('access_token')
-
-            if import_from_settings('OIDC_STORE_ACCESS_TOKEN', False):
-                session['oidc_access_token'] = access_token
-
-            if import_from_settings('OIDC_STORE_ID_TOKEN', False):
-                session['oidc_id_token'] = id_token
-
-            user_response = requests.get(self.OIDC_OP_USER_ENDPOINT,
-                                         headers={
-                                             'Authorization': 'Bearer {0}'.format(access_token)
-                                         },
-                                         verify=import_from_settings('OIDC_VERIFY_SSL', True))
-            user_response.raise_for_status()
-
-            user_info = user_response.json()
-            email = user_info.get('email')
-
-            # email based filtering
-            users = self.filter_users_by_claims(user_info)
-
-            if len(users) == 1:
-                return users[0]
-            elif len(users) > 1:
-                # In the rare case that two user accounts have the same email address,
-                # log and bail. Randomly selecting one seems really wrong.
-                LOGGER.warn('Multiple users with email address %s.', email)
-                return None
-            elif import_from_settings('OIDC_CREATE_USER', True):
-                user = self.create_user(user_info)
-                return user
-            else:
-                LOGGER.debug('Login failed: No user with email %s found, and '
-                             'OIDC_CREATE_USER is False', email)
-                return None
+        verified_id = self.verify_token(id_token, nonce=nonce)
+
+        if verified_id:
+            return self.handle_verified_id(access_token, id_token, verified_id)
+
         return None
 
+    def handle_verified_id(self, access_token, id_token, verified_id):
+        """Get or create the user"""
```

Comment:
The method name and the doc string clash. The fact that it does indeed handle the verified id is an "implementation detail". A more appropriate method name might be "get_or_create_user" and the doc string could be something like 
`"""Returns a User instance if 1 user is found. Creates a user if not found configured to do so. Returns nothing if multiple users are matched."""

Reply:
done